turbopack/crates/turbopack-ecmascript/src/analyzer/side_effects.rs

```diff
@@ -0,0 +1,2337 @@
+                }
+            }
+            Expr::Ident(ident) => {
+                // Check for global pure functions like isNaN, parseInt, etc.
+                // Only consider it pure if the identifier is unresolved (global scope).
+                if ident.ctxt.outer() != self.unresolved_mark {
+                    return false;
+                }
+
+                // O(1) lookup in the global functions set
+                KNOWN_PURE_GLOBAL_FUNCTIONS.contains(ident.sym.as_ref())
+            }
+            _ => false,
+        }
+    }
+
+    /// Check if an expression is a known pure constructor.
+    ///
+    /// These are built-in constructors that create new objects without side effects.
+    /// Only returns true if the identifier is in the global scope (unresolved).
+    /// If it's shadowed by a local variable, we cannot assume it's the built-in constructor.
+    fn is_known_pure_constructor(&self, expr: &Expr) -> bool {
+        match expr {
+            Expr::Ident(ident) => {
+                // Only consider it pure if the identifier is unresolved (global scope).
+                // Check if the identifier's context matches the unresolved mark.
+                if ident.ctxt.outer() != self.unresolved_mark {
+                    return false;
+                }
+
+                // O(1) lookup in the constructors set
+                KNOWN_PURE_CONSTRUCTORS.contains(ident.sym.as_ref())
+            }
+            _ => false,
+        }
+    }
+}
+
+impl<'a> Visit for SideEffectVisitor<'a> {
+    // If we've already found side effects, skip further visitation
+    fn visit_program(&mut self, program: &Program) {
+        if self.has_side_effects {
+            return;
+        }
+        program.visit_children_with(self);
+    }
+
+    fn visit_module(&mut self, module: &Module) {
+        if self.has_side_effects {
+            return;
+        }
+
+        // Only check top-level module items
+        for item in &module.body {
+            if self.has_side_effects {
+                return;
+            }
+            item.visit_with(self);
+        }
+    }
+
+    fn visit_script(&mut self, script: &Script) {
+        if self.has_side_effects {
+            return;
+        }
+
+        // Only check top-level statements
+        for stmt in &script.body {
+            if self.has_side_effects {
+                return;
+            }
+            stmt.visit_with(self);
+        }
+    }
+
+    // Module declarations (imports/exports) need special handling
+    fn visit_module_decl(&mut self, decl: &ModuleDecl) {
+        if self.has_side_effects {
+            return;
+        }
+
+        match decl {
+            // Import statements may have side effects, which could require full graph analysis
+            // Record that to decide if we can upgrade ModuleEvaluationIsSideEffectFree to
+            // SideEffectFree
+            ModuleDecl::Import(_) => {
+                self.has_imports = true;
+            }
```

Comment:
I guess right now all `require`s get handled as containing side-effects, but we should make `require` well-known and do this there too?